core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -113,43 +148,158 @@
+                                          @Nullable TransactionStatus transaction) throws Throwable {
+        Throwable newTransactionFailure = null;
+
+        Collection<NewTransactionListener> listeners = context.getBeansOfType
+            (NewTransactionListener.class).values();
+        for (NewTransactionListener listener : listeners) {
+            try {
+                listener.newTransaction(transaction);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" newTransaction callback failed: "+ t);
+                newTransactionFailure = t;
+            }
+        }
+        if (newTransactionFailure != null){
+            throw newTransactionFailure;
+        }
+    }
+
+    private static void fireAfterRollback(ApplicationContext context,
+                                    PlatformTransactionManager transactionManager,
+                                    @Nullable TransactionStatus transaction) throws Throwable {
+        Throwable afterRollbackFailure = null;
+
+        Collection<AfterRollbackTransactionListener> listeners = context.getBeansOfType
+            (AfterRollbackTransactionListener.class).values();
+        for (AfterRollbackTransactionListener listener : listeners) {
+            try {
+                listener.afterRollback(transaction);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" afterRollback callback failed: "+ t);
+                afterRollbackFailure = t;
+            }
+        }
+        if (afterRollbackFailure != null){
+            throw afterRollbackFailure;
+        }
+    }
+
+    private static void fireBeforeRollback(ApplicationContext context,
+                                           PlatformTransactionManager transactionManager,
+                                           @Nullable TransactionStatus transaction) throws Throwable {
+        Throwable beforeRollbackFailure = null;
+
+        Collection<BeforeRollbackTransactionListener> listeners = context.getBeansOfType
+            (BeforeRollbackTransactionListener.class).values();
+        for (BeforeRollbackTransactionListener listener : listeners) {
+            try {
+                listener.beforeRollback(transaction);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" beforeRollback callback failed: "+ t);
+                beforeRollbackFailure = t;
+            }
+        }
+        if (beforeRollbackFailure != null){
+            throw beforeRollbackFailure;
+        }
+    }
+
+
     private static void doRollback(ApplicationContext context,
                                    PlatformTransactionManager transactionManager,
                                    @Nullable TransactionStatus transaction) {
+        if ( transaction == null || transaction.isCompleted()){
+            // nothing to do
```

Comment:
Is the intention that the code continues?

Reply:
no, reading the previous logic I should have a `return`. Idea was to change the check to be postive, general principle to improve legibility.